src/devices/HCSR04/HCSR04.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Device.Gpio;
+using System.Device.I2c;
+using System.Device.I2c.Drivers;
+using System.Device.Spi;
+using System.Device.Spi.Drivers;
+
+namespace Iot.Device.HCSR04
+{
+    public class Sonar : IDisposable
+    {
+        private readonly int _echo;
+        private readonly int _trigger;
+        private readonly GpioController _controller;
+
+        public Sonar(int triggerPin, int echoPin)
+        {
+            _echo = echoPin;
+            _trigger = triggerPin;
+            _controller = new GpioController();
+
+            _controller.OpenPin(_echo);
+            _controller.OpenPin(_trigger);
+            _controller.SetPinMode(_echo, PinMode.Input);
+            _controller.SetPinMode(_trigger, PinMode.Output);
+        }
+
+        /// <summary>
+        /// Gets the current distance in cm.
+        /// </summary>
+        public double GetDistance()
+        {
+            // Trigger input for 10uS to start ranging
+            _controller.Write(_trigger, PinValue.Low);
+            System.Threading.Thread.Sleep(TimeSpan.FromMilliseconds(1));
+            _controller.Write(_trigger, PinValue.High);
+            System.Threading.Thread.Sleep(TimeSpan.FromMilliseconds(0.01));
+            _controller.Write(_trigger, PinValue.Low);
+
+            // Start timers
+            Stopwatch stopTime = new Stopwatch(); stopTime.Start();
+            Stopwatch startTime = new Stopwatch(); startTime.Start();
+
+            while(_controller.Read(_echo) == PinValue.Low)
+            {
+                startTime.Start();
+            }
+
+            startTime.Stop();
+
+            while(_controller.Read(_echo) == PinValue.High)
```

Comment:
I think something like this would be simpler (make _timer a field in a class):

```csharp
            _timer.Reset();
            // Trigger input for 10uS to start ranging
            _controller.Write(_trigger, PinValue.Low);
            System.Threading.Thread.Sleep(TimeSpan.FromMilliseconds(1));
            _controller.Write(_trigger, PinValue.High);
            System.Threading.Thread.Sleep(TimeSpan.FromMilliseconds(0.01));
            _controller.Write(_trigger, PinValue.Low);

            _timer.Start();

            while(_controller.Read(_echo) == PinValue.Low)
            {
            }

            while(_controller.Read(_echo) == PinValue.High)
            {
            }

            _timer.Stop();
            
            TimeSpan elapsed = _timer.Elapsed;

            // distance = (time / 2) × velocity of sound (34300 cm/s)
            return (double)(elapsed.TotalMilliseconds * 34.3) / 2;
```

In general the code is working for me although I'm getting  around 10cm more than I should which I suspect is related to delays and other stuff

Reply:
Cool, I can test it on my RPi3 today. Are you ok with me adopting the code @krwq?

Reply:
@fredeil sure, feel free to copy/modify or whatever :-)